runsc/container/container.go

```diff
@@ -521,6 +521,28 @@ func Run(conf *config.Config, args Args) (unix.WaitStatus, error) {
 			return 0, fmt.Errorf("starting container: %v", err)
 		}
 	}
+
+	// If we allocate a terminal, forward signals to the sandbox process.
+	// Otherwise, Ctrl+C will terminate this process and its children,
+	// including the terminal.
+	if c.Spec.Process.Terminal {
+		sighandling.StartSignalForwarding(func(signal linux.Signal) {
+			if signal == linux.SIGPIPE || signal == linux.SIGCHLD {
+				return
+			}
+			pid := c.SandboxPid()
+			if pid == -1 {
+				log.Debugf("Failed to forward signal %d to sandbox because its PID is unavailable", signal)
+				return
+			}
+			log.Debugf("Forwarding signal %d to sandbox (pid: %d)", signal, pid)
+			if err := unix.Kill(pid, syscall.Signal(signal)); err != nil {
```

Comment:
This will forward the signal to the sandbox process, which will then forward it to the PID 1 process inside the sandbox because of this SignalForwarding in boot/loader.go:

https://cs.opensource.google/gvisor/gvisor/+/master:runsc/boot/loader.go;l=670;drc=0927892f49b9415bddd5ea4ead2ad53654636521

But we can also call `container.SignalProcess()` which will send the signal directly to the specified PID inside the sandbox directly.  It has the added benefit of checking that the sandbox is running, so it would simplify the code here to just a few lines.
 
https://cs.opensource.google/gvisor/gvisor/+/master:runsc/container/container.go;l=622;drc=e7842860b0b9a8a7367dfa475a0094b8b4abe328

And actually, the `Sandbox.SignalProcess()` takes an additional argument that will signal the *foreground* process group, which I think is what we really want (not just PID 1):

https://cs.opensource.google/gvisor/gvisor/+/master:runsc/sandbox/sandbox.go;l=1089;drc=34ff3ebe055638c57ff59bdf285624db662caef9

I think we should plumb that `fgProcess` argument to `container.SignalProcess()`, and then in your signal handler here we call `c.SignalProcess(signal, 1, true /* fgProcess)`.  Does that work?

Reply:
I also just noticed `container.ForwardSignals()' exists.  Can we make use of that here? 

https://cs.opensource.google/gvisor/gvisor/+/master:runsc/container/container.go;l=636;drc=34ff3ebe055638c57ff59bdf285624db662caef9

Reply:
> I also just noticed `container.ForwardSignals()` exists. Can we make use of that here?
> 
> https://cs.opensource.google/gvisor/gvisor/+/master:runsc/container/container.go;l=636;drc=34ff3ebe055638c57ff59bdf285624db662caef9

You are right. This looks way cleaner now.

https://github.com/google/gvisor/blob/ceea0cb491f48387de3c2cee3ec21e477bfecacd/runsc/container/container.go#L525-L531

Reply:
> And actually, the `Sandbox.SignalProcess()` takes an additional argument that will signal the _foreground_ process group, which I think is what we really want (not just PID 1):
> 
> https://cs.opensource.google/gvisor/gvisor/+/master:runsc/sandbox/sandbox.go;l=1089;drc=34ff3ebe055638c57ff59bdf285624db662caef9
> 
> I think we should plumb that `fgProcess` argument to `container.SignalProcess()`, and then in your signal handler here we call `c.SignalProcess(signal, 1, true /* fgProcess */)`. Does that work?

Instead of sending the signal to PID 1, the signal is sent to PID 0, the container init process. This boils down to calling `signalForegrondProcessGroup` with a `tgid` of `0` in `loader.go`. Since a terminal is attached, the function will first resolve the tty:

https://github.com/google/gvisor/blob/8184fa1db0faaa9cf742a6c094f40e5442635bf1/runsc/boot/loader.go#L1307

From this tty, the foreground process group is resolved, which includes the interactive shell that was started inside the container.

https://github.com/google/gvisor/blob/8184fa1db0faaa9cf742a6c094f40e5442635bf1/runsc/boot/loader.go#L1307

In the case of `exec`, we know the PID of the process we just started inside the sandbox from the `Execute` function. Then, the process group of this PID is used.
https://github.com/google/gvisor/blob/ceea0cb491f48387de3c2cee3ec21e477bfecacd/runsc/cmd/exec.go#L155-L165